core/trino-main/src/main/java/io/trino/operator/join/unspilled/PartitionedLookupSource.java

```diff
@@ -144,6 +144,58 @@ public long getJoinPosition(int position, Page hashChannelsPage, Page allChannel
         return encodePartitionedJoinPosition(partition, toIntExact(joinPosition));
     }
 
+    @Override
+    public void getJoinPosition(int[] positions, Page hashChannelsPage, Page allChannelsPage, long[] rawHashes, long[] result)
+    {
+        int positionCount = positions.length;
+        int partitionCount = partitionGenerator.getPartitionCount();
+
+        int[] partitions = new int[positionCount];
+        int[] partitionPositionsCount = new int[partitionCount];
+
+        // Get the partitions for every position and calculate the size of every partition
+        for (int i = 0; i < positionCount; i++) {
+            int partition = partitionGenerator.getPartition(rawHashes[positions[i]]);
+            partitions[i] = partition;
+            partitionPositionsCount[partition]++;
+        }
+
+        int[][] positionsPerPartition = new int[partitionCount][];
+        for (int partition = 0; partition < partitionCount; partition++) {
+            positionsPerPartition[partition] = new int[partitionPositionsCount[partition]];
+        }
+
+        // Split input positions into partitions
+        int[] positionsPerPartitionCount = new int[partitionCount];
+        for (int i = 0; i < positionCount; i++) {
+            int partition = partitions[i];
+            positionsPerPartition[partition][positionsPerPartitionCount[partition]] = positions[i];
+            positionsPerPartitionCount[partition]++;
+        }
+
+        // Delegate partitioned positions to designated lookup sources
+        for (int partition = 0; partition < partitionCount; partition++) {
+            lookupSources[partition].getJoinPosition(positionsPerPartition[partition], hashChannelsPage, allChannelsPage, rawHashes, result);
+        }
+
+        for (int i = 0; i < positionCount; i++) {
+            int partition = partitions[i];
+            result[positions[i]] = encodePartitionedJoinPosition(partition, (int) result[positions[i]]);
+        }
+    }
+
+    @Override
+    public void getJoinPosition(int[] positions, Page hashChannelsPage, Page allChannelsPage, long[] result)
+    {
+        int positionCount = positions.length;
+        long[] rawHashes = new long[result.length];
```

Comment:
nit: do we waste memory here potentially? Can `positions.length <<< result.length` (much smaller)?

Reply:
In this exact place, `positions` array will hold all non-null values. So `result.length` is significantly bigger than `positions.length` only when there is a lot of nulls.